components/search/src/types.rs

```diff
@@ -0,0 +1,197 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// The list of possible application names that are currently supported.
+#[derive(Debug, uniffi::Enum)]
+pub enum SearchApplicationName {
+    Firefox = 1,
+    FirefoxAndroid = 2,
+    FocusAndroid = 3,
+    FirefoxIOS = 4,
+    FocusIOS = 5,
+}
+
+impl SearchApplicationName {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            SearchApplicationName::Firefox => "firefox",
+            SearchApplicationName::FirefoxAndroid => "firefox-android",
+            SearchApplicationName::FocusAndroid => "focus-android",
+            SearchApplicationName::FirefoxIOS => "firefox-ios",
+            SearchApplicationName::FocusIOS => "focus-ios",
+        }
+    }
+}
+
+/// The list of possible update channels for a user's build.
+/// Use `default` for a self-build or an unknown channel.
+#[derive(Debug, uniffi::Enum)]
+pub enum SearchUpdateChannel {
+    Default,
+    Nightly,
+    Aurora,
+    Beta,
+    Release,
+    ESR,
+}
+
+impl SearchUpdateChannel {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            SearchUpdateChannel::Default => "default",
+            SearchUpdateChannel::Nightly => "nightly",
+            SearchUpdateChannel::Aurora => "aurora",
+            SearchUpdateChannel::Beta => "beta",
+            SearchUpdateChannel::Release => "release",
+            SearchUpdateChannel::ESR => "esr",
+        }
+    }
+}
+
+/// The user's environment that is used for filtering the search configuration.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchUserEnvironment {
+    /// The current locale of the application that the user is using.
+    pub locale: String,
+
+    /// The home region that the user is currently identified as being within.
+    /// On desktop & android there is a 14 day lag after detecting a region
+    /// change before the home region changes. TBD: iOS?
+    pub region: String,
+
+    /// The update channel of the user's build.
+    pub update_channel: SearchUpdateChannel,
+
+    /// The distribution id for the user's build.
+    pub distribution_id: String,
+
+    /// The search related experiment id that the user is included within. On
+    /// desktop this is the `searchConfiguration.experiment` variable.
+    pub experiment: String,
+
+    /// The application name that the user is using.
+    pub app_name: SearchApplicationName,
+
+    /// The application version that the user is using.
+    pub version: String,
+}
+
+/// Parameter definitions for search engine URLs. The name property is always
+/// specified, along with one of value, experiment_config or search_access_point.
+#[derive(Debug, uniffi::Record)]
+pub struct SearchUrlParam {
+    /// The name of the parameter in the url.
+    pub name: String,
+    /// The parameter value, this may be a static value, or additionally contain
+    /// a parameter replacement, e.g. `{inputEncoding}`. For the partner code
```

Comment:
Idle thought: Would you see us moving parameter replacements into the Rust layer eventually, too (and having it return the "cooked" URLs, with replacements interpolated), or leaving that up to the consuming Desktop / Android / iOS code?

Reply:
I certainly have a background thought for the long term that we could potentially completely rewrite the search service in Rust. I think it is also possible that we could have `SearchEngineUrl` be a more featured URL that returns the filled-in URLs. I'm also trying to avoid a bit of scope creep 😉, though we'll keep it in mind.